File: src/com/easy/MergeTwoBinaryTrees_617.java
/**
 * 
 */
package com.easy;

import java.util.LinkedList;
import java.util.Queue;

/**
 * @author Chris Lin
 * @version 1.1
 */
public class MergeTwoBinaryTrees_617 {

    // Use recursive
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 69.01%
    // Memory Usage: 40.1 MB, less than 100.00%
    public TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
        if (t1 == null && t2 == null) {
            return null;
        }
        if (t1 == null) {
            return t2;
        }
        if (t2 == null) {
            return t1;
        }
        TreeNode mMergeTree = new TreeNode(t1.val + t2.val);
        mMergeTree.left = mergeTrees(t1.left, t2.left);
        mMergeTree.right = mergeTrees(t1.right, t2.right);

        return mMergeTree;
    }

    // Use iterative
    // Time Complexity: O(n)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 69.01%
    // Memory Usage: 39.7 MB, less than 100.00%
    public TreeNode mergeTrees2(TreeNode t1, TreeNode t2) {
        if (t1 == null) {
            return t2;
        }
        if (t2 == null) {
            return t1;
        }
        Queue<TreeNode[]> mQueue = new LinkedList<>();

        mQueue.add(new TreeNode[] { t1, t2 });

        while (!mQueue.isEmpty()) {
            TreeNode[] mNode = mQueue.poll();

            if (mNode[1] == null) {
                continue;
            }

            mNode[0].val += mNode[1].val;
            if (mNode[0].left == null) {
                mNode[0].left = mNode[1].left;
            } else {
                mQueue.add(new TreeNode[] { mNode[0].left, mNode[1].left });
            }
            if (mNode[0].right == null) {
                mNode[0].right = mNode[1].right;
            } else {
                mQueue.add(new TreeNode[] { mNode[0].right, mNode[1].right });
            }
        }
        return t1;
    }
}

File: Golang_src/easy/mergeTwoBinaryTrees_617.go
package easy

// Use DFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 32 ms, faster than 30.72%
// Memory Usage: 8.6 MB, less than 92.81%
func mergeTrees(t1 *TreeNode, t2 *TreeNode) *TreeNode {
	var result *TreeNode

	if t1 == nil && t2 == nil {
		return nil
	}

	tmp1, tmp2 := 0, 0
	isNil1, isNil2 := true, true
	if t1 != nil {
		tmp1 = t1.Val
		isNil1 = false
	}
	if t2 != nil {
		tmp2 = t2.Val
		isNil2 = false
	}
	node := &TreeNode{Val: tmp1 + tmp2}
	result = node

	if isNil1 {
		result.Left = mergeTrees(nil, t2.Left)
		result.Right = mergeTrees(nil, t2.Right)
	} else if isNil2 {
		result.Left = mergeTrees(t1.Left, nil)
		result.Right = mergeTrees(t1.Right, nil)
	} else {
		result.Left = mergeTrees(t1.Left, t2.Left)
		result.Right = mergeTrees(t1.Right, t2.Right)
	}

	return result
}

// Optimize DFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 28 ms, faster than 90.20%
// Memory Usage: 8.8 MB, less than 92.81%
func mergeTrees2(t1 *TreeNode, t2 *TreeNode) *TreeNode {
	if t1 == nil {
		return t2
	}
	if t2 == nil {
		return t1
	}
	t1.Val += t2.Val
	t1.Left = mergeTrees2(t1.Left, t2.Left)
	t1.Right = mergeTrees2(t1.Right, t2.Right)

	return t1
}
